docs/design/specs/runtime-async.md

```diff
@@ -0,0 +1,71 @@
+
+This document is a draft of changes to ECMA-335 for the "runtime async" feature. When the feature is officially supported, it can be merged into the final ECMA-335 augments document.
+
+# Runtime-async
+
+Async is currently a feature implemented by various .NET languages as a compiler rewrite to support methods that can "yield" control back to their caller at specific "suspension" points. While effective, it's believed that implementation directly in the .NET runtime can provide improvements, especially in performance.
+
+## Spec modifications
+
+These are proposed modifications to the ECMA-335 specification for runtime-async.
+
+### I.8.4.5 Sync and Async Methods
+
+Methods may be either 'sync' or 'async'. Async methods have a special signature encoding, described in [### I.8.6.1.5 Method signatures].
+
+Sync methods are all other methods.
+
+Unlike sync methods, async methods support suspension. Suspension allows async methods to yield control flow back to their caller at certain well-defined suspension points, and resume execution of the remaining method at a later time or location, potentially on another thread.
+
+Async methods support the following suspension points:
+
+* Calling another async method. No special instructions need to be provided. If the callee suspends, the caller will suspend as well.
+* Using a library in the .NET corelib to "await" an "INotifyCompletion" type. The signatures of these methods shall be:
+  ```C#
+  // public static async2 Task AwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion
+  public static void modopt([System.Runtime]System.Threading.Tasks.Task) AwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion
+  // public static async2 Task UnsafeAwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : ICriticalNotifyCompletion
+  public static void modopt([System.Runtime]System.Threading.Tasks.Task) UnsafeAwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : ICriticalNotifyCompletion
+  ```
+
+Each of the above methods will have semantics analogous to the current AsyncTaskMethodBuilder.AwaitOnCompleted/AwaitUnsafeOnCompleted methods. After calling this method, in can be presumed that the task has completed.
+
+Async methods have the following restrictions:
```

Comment:
What about byref-like types? We should make sure that the rules work for https://github.com/dotnet/csharplang/blob/main/proposals/ref-unsafe-in-iterators-async.md

Reply:
Given this restriction is only going to be lifted in .NET 9, I think it would be OK to leave this support out of v1.

Reply:
What is Roslyn going to do when it runs into this and targeting runtime that supports async v2? Is it going to silently fallback to async v1? It does not sound right to me.

I think the support for this may be actually a good justification for introducing annotations of non-escaping / escaping locals: Non-escaping locals work just like locals today. Escaping locals cannot by ref or byref like types, taking their address is ref and not an unmanaged pointer, etc.

Reply:
> I think the support for this may be actually a good justification for introducing annotations of non-escaping / escaping locals: Non-escaping locals work just like locals today. Escaping locals cannot be ref or byref like types, taking their address is a byref and not an unmanaged pointer, etc.

I am not sure there is a good reason for us to have these annotations unless/until we also decide to support byrefs that are live across suspension. If live byrefs are not allowed then liveness analysis can treat any address exposed local to no longer have escaped at any suspension point and only save it if there are actual visible uses/defs after the suspension point (which may include taking the address again).

I think the support for https://github.com/dotnet/csharplang/blob/main/proposals/ref-unsafe-in-iterators-async.md comes down to how we reword the "Transient pointers" section to describe the address of locals. I pointed at it here: https://github.com/dotnet/runtime/pull/104063#discussion_r1656968365

Reply:
> I am not sure there is a good reason for us to have these annotations unless/until we also decide to support byrefs that are live across suspension

To make it possible to compute types on IL stack without the full liveness analysis and to avoid changing the type of ldarga from unmanaged pointer to managed pointer for locals that are never live across transitions. I agree that these annotations are not strictly required to make the system work.

> support byrefs that are live across suspension.

If we took a bet on liveness analysis, I think that this can be done by the liveness analysis too.

Reply:
> To make it possible to compute types on IL stack without the full liveness analysis and to avoid changing the type of ldarga from unmanaged pointer to managed pointer for locals that are never live across transitions.

Do you mean spec wise or implementation wise? Spec wise these are all managed pointers already, and I don't see a good reason to change that. I think we should just document that they reside in unmanaged memory that can be used without pinning, but that their address may move across suspension points. Then it comes down to the fact that we do not preserve values with byrefs/structs with byrefs across suspension points.

Implementation wise I don't think there will be a difference for these locals in the JIT, async or not. The liveness analysis is only about computing the state to capture at each suspension point.

> If we took a bet on liveness analysis, I think that this can be done by the liveness analysis too.

When byrefs are allowed to be live across suspension points Roslyn can do a better job than the JIT can, because it has a richer model about what may escape. My example from the internal document: if the JIT sees a call to an instance method on a struct, then it is going to consider that struct to be exposed.
If there is also a byref live across a suspension point, then that byref may point to the exposed struct. In effect the JIT needs to consider both the exposed struct and byref to be live across the suspension point.

On the other hand Roslyn only needs to consider the instance as escaping if it also has `[UnscopedRef]` on it; otherwise it knows that the address of the instance was not actually taken once the method call returns. So it can annotate that the struct local is not actually the target of any byrefs across suspension points.

I _think_ when byrefs are not allowed across suspension points the code generators can do a job that in practice is as good as Roslyn, via the liveness analysis. Then it comes down to whether we want to have a conservatively larger liveness set in tier0/MinOpts, or whether we can pay for the liveness analysis even during tier0/MinOpts. I think we can -- in my experience these kind of throughput questions usually come down to engineering within the JIT.

Reply:
I was initially leaving "unhoisted ref-locals" out of the spec to keep the implementation simpler. But it sounds like the implementors are fine with the additional work. So I'll put it in. :smile:

Reply:
> these annotations are not strictly required to make the system work

We would really really like to have the annotations for the tier0 interpreter.  We would like to avoid:
1. having to do our own analysis in tier0 to determine which locals need to be preserved in the closure. That will slow down startup.
2. having tier0 and optimized versions of an async2 function use different closure representations (for example tier0 that conservatively saves everything and optimized code that only saves a subset)

Reply:
I've also put in the statement about marking hoisted variables.

Reply:
> Do you mean spec wise or implementation wise? Spec wise these are all managed pointers already, and I don't see a good reason to change that. 

They are managed pointers with extra promise about being implicitly pinned. My primary concern was about ensuring that there are no subtle mismatches in how different parts of the system interpret liveness. Let's consider this hypothetical example:

```
void M()
{
    
    {
       int x;
       foo(&x);
    }

    await SomeAsyncMethod();

    {
       int x;
       foo(&x);
    }
}
```

This should get compiled into IL with two local variables, so far so good. Now, some optimization pass observes that there are two `x` variables with same type and disjoint lifetimes and decides to merge them into one. Oops, it just introduced a subtle GC hole. The merged variable liveness is crossing suspension point and the address is no longer pinned.

> whether we can pay for the liveness analysis even during tier0/MinOpts. I think we can -- in my experience these kind of throughput questions usually come down to engineering within the JIT

I agree that cost of the analysis should not be our primary motivation for introducing these annotations. It should not be prohibitively expensive to do the liveness analysis even in tier 0 or interpreter.

> "unhoisted ref-locals" out of the spec to keep the implementation simpler

I do not think allowing byrefs across suspension points should be part of the initial design. The initial design should be oriented around performance and compatibility. Support for byrefs crossing suspension points is not that.

Reply:
> The merged variable liveness is crossing suspension point and the address is no longer pinned.

This is further motivation to me that we should specify that locals live in static memory in-between suspension points, but that they may move across suspension points. I do not see a problem with your example case under this specification.

> I was initially leaving "unhoisted ref-locals" out of the spec to keep the implementation simpler. But it sounds like the implementors are fine with the additional work. So I'll put it in. 😄

I would prefer the opposite: no annotations and it should be left up to the code generator to compute a sufficient live set to allow for resumption with the right semantics. When optimizing the code generators (at least RyuJIT) will create their own internal locals that is going to require this analysis anyway.

The current spec seems a bit odd to me; requiring non-marked locals to be reinitialized on resumption has strange implications. For example, what is the semantics of a program `x = 5; await Foo(); Use(x)` where `x` is unmarked when `Foo` did not suspend? Is `x = 5` or `x = 0`?

> having tier0 and optimized versions of an async2 function use different closure representations (for example tier0 that conservatively saves everything and optimized code that only saves a subset)

@lambdageek Does this mean that you expect Mono to create no internal locals that are live across suspension points, even when optimizing? I do not see this as a possible strategy for RyuJIT when we want performance close to "normal" straight-line code in the synchronous cases.

Reply:
> This is further motivation to me that we should specify that locals live in static memory in-between suspension points, but that they may move across suspension points. I do not see a problem with your example case under this specification.

How would this specification be implemented in practice? I can think of: Keep the address taken locals on the stack and copy their values between the stack and the state machine as necessary (that is fine with me); or pin the state machine (this sounds problematic at scale).

Reply:
> How would this specification be implemented in practice? I can think of: Keep the address taken locals on the stack and copy their values between the stack and the state machine as necessary (that is fine with me); or pin the state machine (this sounds problematic at scale).

In the JIT state machine prototype we indeed do the former. On resumption the live set is copied from the continuation back to registers/the stack frame, and control resumes in the same code that was running before suspension.

I believe the unwinder-based prototype does similar copying for 'tasklets' before they are executed, but I could be wrong about that.

Reply:
> I expect that when the interpreter needs to create locals for its own use we will know from context whether we expect the local to live across suspension points (or at least we will conservatively know which locals definitely don't need to live across suspension points).

What I mean is that it is almost inevitable that optimized code has a different closure representation than the unoptimized code, at least if you are optimizing the code as if it was normal synchronous code. Things like inlining, constant propagation, copy propagation etc. are going to result in less live state across suspension points while things like CSE, promotion, hoisting are going to result in more live state across the suspension points. Trying to match the closure between the unoptimized and optimized versions feels like a non-starter to me.

Reply:
@lambdageek With interpreter, tiering up during method execution can only happen at certain patchpoint locations where the IL stack is empty. The method execution state is thus fully identified by the values of IL args and locals. For optimized code, after a patchpoint, we insert some IR markers for IL vars that are live at that point, that behave as a redefinition for those vars. This forces following uses of the var to obtain the value from the IL locals space, in case the var value comes from code executed in tier0. We don't do any copying of the state when tiering up, the optimized code just works naturally with the values in the IL locals/args. Stackalloc memory is allocated separately from the interp stack where the IL args/locals live.

Reply:
I am going to keep the annotations for now. It's easier to plan for them and not need them then the reverse.